Add ARM option to building packages on Jenkins

jclarkeSTFC opened the pull request:
This adds the option to build a native package for Arm Macs. It will be listed as a new option called `osx-arm64`, with the existing option remaining as `osx-64`.

The `osx-arm64` platform can only run on Arm macOS machines, and it will compile native Arm code instead of cross-compiling for `x86_64`. The platform is used for specifying the cmake preset, which is why I added the `osx-arm64-ci` preset.

Needed for #38973 
Fixes #38995 

In various places the desired platform is passed as an argument. We cannot simply use the native platform, because any ARM build machine could be either targeting `osx-64` or `osx-arm64`, so we need to know what was selected.

I've simplified some code in `mamba-utils`, and deleted the bit that was forcing `osx-64` on Macs. 

When archiving the environment logs we get the platform from the configuration of the current conda environment. This is because that script is called from the build script in the conda recipe, which doesn't know about options specified in Jenkins, so it seemed like the best way of getting that information.

The `jemalloc` library is pinned to version `5.2.0` for Linux, but `5.2.1` is the first version with an Arm build. Luckily we don't need `jemalloc` on macOS anyway.

~~Version `1.0.0b15` of `quickbayes` does not have an ARM build, so the `pip install` when creating the standalone will fail in that case. Version `1.0.1b0` is a `noarch` package, so works on ARM, but it has a pin on `numpy<2`, so if we `pip install` that
then `numpy` in the standalone will be downgraded from 2.* to 1.*. Hence we do not install `quickbayes` into the macOS ARM standalone package.~~
`quickbayes` is now a conda package, and is a `noarch` package, so this isn't a problem any more.

Several tests (especially anything that involves fitting) generate different numbers on ARM machines. In some places I've adjusted the tolerance, but in others (especially some system tests), I've just skipped the test on ARM. 

### To test:
- On an Arm machine, check that the package runs okay, and is running natively, not under Rosetta.
- On an Arm machine, check that the old (Intel) package runs okay, and is running under Rosetta.
- Try running the `Core_Team_test_pipeline` with the `osx-arm64` option and the `osx-64` option.
- Build and test standalone packages on all platforms.

Standalone:

https://builds.mantidproject.org/job/Core_Team_test_pipeline/247/artifact/

- [x] Windows @jhaigh0 
- [x] macOS x86_64 on x86_64 @SilkeSchomann 
- [x] macOS x86_64 on ARM @adriazalvarez 
- [x] macOS ARM on ARM @adriazalvarez 
- [x] Linux @MohamedAlmaki 

Conda:

`conda install jclarkestfc/label/armtest::mantidworkbench -c mantid`

- [x] Windows @jhaigh0 
- [x] macOS x86_64 @SilkeSchomann 
- [x] macOS ARM @adriazalvarez 
- [x] Linux @thomashampson 

When testing on an ARM machine, opening Activity Monitor will show if the process is 'Intel' or 'Apple', i.e. x86 or ARM.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Add Arm option to building packages on Jenkins

The osx-arm64 platform can only run on Arm macOS machines, and it will
compile native Arm code instead of cross-compiling for x86_64. The
platform is used for specifying the cmake preset, which is why I added
the osx-arm64-ci preset.

jclarkeSTFC committed:
Pass optional platform argument to build scripts

We need the option to be able to specify the platform so Arm Macs can
either build for osx-64 or osx-arm64. In conda-buildscript we take
the platform from the cmake preset, which should then work with
existing CI. To activate Arm we then just use the osx-arm64-ci preset.

jclarkeSTFC committed:
Add platform argument to setup_mamba

Allow for passing in the target platform, e.g. linux-64 to the
setup_mamba method. The `clean` parameter is now no longer optional
to simplify matters. Where setup_mamba was called without the clean
parameter it now specifies `false` (the old default).

jclarkeSTFC committed:
Only pin jemalloc on Linux

jclarkeSTFC committed:
Remove macOS deployment target for Arm

Using the old SDK results in a linker error:

ld: warning: ignoring file /opt/MacOSX10.10.sdk/usr/lib/libSystem.tbd,
missing required architecture arm64 in file /opt/MacOSX10.10.sdk/usr/
lib/libSystem.tbd (2 slices)

macOS 10.10 was released in 2014. The first version of macOS with Arm
support was 11 (Big Sur), released in 2020.

jclarkeSTFC committed:
Use the conda subdir to get the target platform

The archive_env_logs script needs to know the platform in order to
extract the log files.

jclarkeSTFC committed:
Pass platform through to standalone

We need the target platform in the script that creates the dmg for
macOS, since it could be either osx-64 or osx-arm64. Previously it was
always osx-64 so could be hard-coded.

I also added the platform name to the dmg file name to avoid a clash
when making both the osx-64 and osx-arm64 standalones.

jclarkeSTFC commented:
The standalone scripts install `quickbayes`, and the version of `quickbayes` specified in the scripts does not have an Arm build. Therefore we need #39014 to be merged, since that updates `quickbayes` to a version that is architecture independent.

jclarkeSTFC commented:
This is blocked because the version of `quickbayes` we `pip install` at the end of the packaging step does not have an Arm build. #39014 will update `quickbayes` to a version that is `noarch` so should be no problem. `GOFit` is an (optional, in latest version) dependency of `quickbayes`, and now there is now an Arm version of that available after I added it to the migration list on conda-forge.

I've also made a PR https://github.com/conda-forge/staged-recipes/pull/29358 to add `quickbayes` to conda-forge, so once that is done we can remove the `pip install` from the script. That way any dependencies will be much easier to spot in the future.

jclarkeSTFC commented:
Package build all platforms: [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/186/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/186/)

jclarkeSTFC committed:
Use variable for log directory and echo everything

jclarkeSTFC committed:
Make sure platform is correct for packaging env

When archiving the environments, we get the platform from the conda
environment, so we need it to be the correct platform. This only
currently applies when building for osx-64 an a osx-arm64 machine.

jclarkeSTFC commented:
Package build all platforms: [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/192/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/192/)

jclarkeSTFC committed:
Add build_and_test option for osx-arm64

jclarkeSTFC commented:
Build and test, then package build, all platforms: [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/195/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/195/)

jclarkeSTFC commented:
Turns out that the ARM `quasielasticbayes` package has `x86_64` binaries, so https://github.com/conda-forge/quasielasticbayes-feedstock/pull/8 will fix that once it's merged in.

jclarkeSTFC committed:
These comparisons with zero should be using a tolerance

jclarkeSTFC committed:
Use tolerances when comparing floats

Not a good idea to compare floating point numbers with either a zero
tolerance, or even a tolerance smaller than machine precision. Once
you start doing operations on a number you can't expect the result
to be absolutely identical when different CPU instruction sets are
used.

jclarkeSTFC commented:
Build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/198/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/198/)

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/207/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/207/)

jclarkeSTFC commented:
pre-commit.ci run

jclarkeSTFC commented:
#39126 is required to fix a problem with loading Nexus files.

sf1919 commented:
As you need #39126 to go in before this one I won't re-run the tests right now. Please note though that `11501` has failed due to the dependency issue with `pydantic`

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/211/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/211/)

jclarkeSTFC committed:
Use significant figures in test comparison

The deltas are arbitrary and in some cases don't make sense, since the
number that was typed in was several significant figures more
accurate than the delta.

jclarkeSTFC committed:
Relax tolerances in several tests

This avoids false negatives when running the tests on different
architectures. EvaluateMDFunction is a deprecated algorithm.

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/212/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/212/)

jclarkeSTFC commented:
#39205 is required for this PR

jclarkeSTFC commented:
#39215 is required for this PR

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/222/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/222/)

jclarkeSTFC committed:
Add tolerance to d-spacing perpendcular test

This test sends the min and max values on a round trip, then for the
min value for d-spacing is a very small amount less than the original
min on ARM processors. So use a tolerance for that check like in
several other places.

jclarkeSTFC committed:
Ignore some tests on ARM

Fitting can give different results on ARM, so for these tests the
results are different. The spline tests were failing because the final
x point was causing numerical problems, so I reduced the range very
slightly.

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/228/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/228/)

jclarkeSTFC committed:
Skip tests on ARM

The ILL tests time out, while the BayesQuasiTest is constantly giving
different results with every change, e.g. ARM, alma9, package updates,
etc.

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/229/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/229/)

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/230/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/230/)

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/231/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/231/)

jclarkeSTFC committed:
Ignore system tests with different numbers on ARM

jclarkeSTFC committed:
Use tolerance in POLDIFitPeaks2DTest comparison

jclarkeSTFC committed:
Adjust tolerance on ARM

jclarkeSTFC commented:
ARM build and test [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/234/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/234/)

jclarkeSTFC commented:
Build, test, package all platforms [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/235/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/235/)

jclarkeSTFC committed:
Use target_platform for environment logs

This environment variable should have the correct platform. On macOS we
can't just get harware information because we need to know if we are
cross-compiling for x86_64 or not.

jclarkeSTFC commented:
Package all platforms [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/240/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/240/)

jclarkeSTFC commented:
Package osx-64 [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/241/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/241/)

jclarkeSTFC commented:
Package all platforms [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/242/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/242/)

jhaigh0 reviewed (commented):
Can install, open, load some data and plot with the Windows installer 👍

jhaigh0 commented on conda/recipes/archive_env_logs.sh:
```diff
@@ -1,24 +1,19 @@
 #!/usr/bin/env bash
 
+set -x
+
```
was this for debugging?

adriazalvarez commented:
Both standalone installer (mac x86_64 and arm) work as expected on ARM device. Tried opening few interfaces, load data, plot and fit. Everything seems in order. The same goes for conda install.

MohamedAlmaki commented:
I get this error when trying to open Ubuntu. I have tried previous stable and unstable builds, and they are working fin,e so it is probably not a problem with my local config
![image](https://github.com/user-attachments/assets/ad133703-8036-4121-81f8-0ebfad45bcda)

jclarkeSTFC commented:
> I get this error when opening it on Ubuntu (no stacktrace, but it looks like it couldn't find a Python installation). I have tried previous stable and unstable builds, and they are working fine, so it is probably not a problem with my local config ![image](https://private-user-images.githubusercontent.com/44416400/441247361-ad133703-8036-4121-81f8-0ebfad45bcda.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDY2MjI4MjksIm5iZiI6MTc0NjYyMjUyOSwicGF0aCI6Ii80NDQxNjQwMC80NDEyNDczNjEtYWQxMzM3MDMtODAzNi00MTIxLTgxZjgtMGViZmFkNDViY2RhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTA3VDEyNTUyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1YWFkMjBiYmFlMDVmNDA2N2RmMWIyNWQwYjg4ZDQ3OTdlMDM3MTljZjViYmUyNWVlMDRjNzk2MGM5NDIzYjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hPk4zF2UMqpgsgScyYkxRrjMDQUb8gHTCrTuWiqAO-M)

This is fixed by https://github.com/mantidproject/mantid/pull/39239, I need to rebase and rebuild the packages

jclarkeSTFC committed:
Remove echoing in archive_env_logs.sh

jclarkeSTFC commented:
Package all platforms [![Build Status](https://builds.mantidproject.org/job/Core_Team_test_pipeline/247/badge/icon)](https://builds.mantidproject.org/job/Core_Team_test_pipeline/247/)

Rebased to include fix for launching, see #39239

SilkeSchomann commented:
Both the standalone and the conda package for macOS x86_64 on x86_64 work as expected. I've tried out loading data, plotting, fitting and a few algorithms.

MohamedAlmaki commented:
The new version is now working well on my Linux machine. I have tried loading, plotting, and fitting data, using various interfaces and algorithms.

I noticed an error in the first line of the log, however, it is not a regression since it also appears in previous versions.  Looks like a non-breaking issue, but might be worth writing an issue for it for investigation:

```
Fontconfig error: Cannot load default config file: No such file: (null)
```

MBartkowiakSTFC commented:
I did some basic manual testing of the ARM build on an M1 Mac. 

Minor glitches:
1. MacOS could not verify the app, and I had to go through the (usual) procedure of overriding the security restrictions.
2. The first time I loaded a data file (measured on POLARIS), it took over 80 seconds. I guess Mantid was checking/updating instrument definitions? The second time the same file loaded in under a second.
3. Mantid Help showed me a window with "Missing page - qthelp://org.mantidproject/doc/index.html" the first time I tried it. Some seconds later I tried again, and it worked.
4. Default path for the Save/Open Project dialogs seems to be "/".
5. Every time I start MantidWorkbench, a bash process appears in my Activity Monitor, which is still running in "intel" mode, i.e. via Rosetta. So, some component of the entire workflow is still not running natively. Not sure if it matters.

Things that worked:
- The GUI started normally, and was shown to be running natively ("apple" and not "intel").
- Instrument view showed a realistic instrument geometry. Grouping spectra in 3D view and exporting the sum to a new workspace worked too.
- Accessing a workspace in IPython and editing the axis unit and data arrays worked fine.
- DGSPlanner and PyChopGUI worked, and produced reasonable values.
- The plotting and fitting interface worked, and produced a result (but I did not have a good test case ready to check the details).
- Saving a workspace history to script and running the script worked fine.

Things I did not check:
- I did not manage to create an MD workspace out of my data files, so I could not try any part of the GUI that deals with MD workspace plotting and slicing.
I thought I should be able to run ConvertToDiffractionMDWorkspace on my workspace (diffraction data from Polaris), but I got this every time:
```
ConvertToDiffractionMDWorkspace started
Error in execution of algorithm PreprocessDetectorsToMD:
Cannot calculate an angle when one of the vectors has zero length.
Error in execution of algorithm ConvertToMDMinMaxLocal:
Cannot calculate an angle when one of the vectors has zero length.
Error in execution of algorithm ConvertToDiffractionMDWorkspace:
Cannot calculate an angle when one of the vectors has zero length.
```
I have not used Mantid in a few years, so the fact that I could not get this to run may have nothing to do with this PR.

jclarkeSTFC commented:
Novel issues have been written up separately, #39325, #39326
